### ISSUE TITLE
Dynamicuniformbuffer incorrect usage

### ISSUE THREAD

**Issue by BobLChen** [SaschaWillems/Vulkan#604]
https://github.com/SaschaWillems/Vulkan/blob/master/examples/dynamicuniformbuffer/dynamicuniformbuffer.cpp#L328

VkDescriptorBufferInfo.range equal vertexshader.uboInstance size, not the whole buffer size.

https://github.com/SaschaWillems/Vulkan/pull/601/commits/db923241322ff9be3faf7e6cbd6ee64ffcc431a7

**Comment by SaschaWillems** [comment]
Can you add some details? What's actually wrong with that line? The sample works fine on a range of implementations and validation is clean too.

The size for that buffer is simply set to ```VK_WHOLE_SIZE```.

**Comment by BobLChen** [comment]
The validation error will appear if you set OBJECT_INSTANCES to [1250](https://github.com/SaschaWillems/Vulkan/blob/master/examples/dynamicuniformbuffer/dynamicuniformbuffer.cpp#L39)

from
```c++
#define OBJECT_INSTANCES 125
```
to
```c++
#define OBJECT_INSTANCES 1250
```

**Comment by SaschaWillems** [comment]
And what's the actual error you get? I still think my code is correct, and that you instead run into some hardware limitations.

**Comment by BobLChen** [comment]
Error:
```c++
ERROR: [Validation] Code 0 :  [ VUID-VkWriteDescriptorSet-descriptorType-00332 ] Object: 0x34 (Type = 23) | vkUpdateDescriptorSets() failed write update validation for Descriptor Set 0x34 with error: Write update to descriptor in set 0000000000000034 binding #1 failed with error message: Attempted write update to buffer descriptor failed due to: VkDescriptorBufferInfo range is VK_WHOLE_SIZE but effective range (320000) is greater than this device's maxUniformBufferRange (65536). The Vulkan spec states: If descriptorType is VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER or VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC, the range member of each element of pBufferInfo, or the effective range if range is VK_WHOLE_SIZE, must be less than or equal to VkPhysicalDeviceLimits::maxUniformBufferRange (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkWriteDescriptorSet-descriptorType-00332)
```

The range member of VkDescriptorBufferInfo can't set to **VK_WHOLE_SIZE**, it must be object's uniform size. A normal scenes can contain more than 1250 objects.

change from
```c++
VkDescriptorBufferInfo.range = VK_WHOLE_SIZE;
```
to
```c++
VkDescriptorBufferInfo.range = sizeof(glm::mat4);
```

the corrected code should be
```c++
void setupDescriptorSet()
{
	VkDescriptorSetAllocateInfo allocInfo =
		vks::initializers::descriptorSetAllocateInfo(
			descriptorPool,
			&descriptorSetLayout,
			1);

	VK_CHECK_RESULT(vkAllocateDescriptorSets(device, &allocInfo, &descriptorSet));

	VkDescriptorBufferInfo dynamicBufferInfo;
	dynamicBufferInfo.buffer = uniformBuffers.dynamic.buffer;
	dynamicBufferInfo.offset = 0;
	dynamicBufferInfo.range  = sizeof(glm::mat4);

	std::vector<VkWriteDescriptorSet> writeDescriptorSets = {
		// Binding 0 : Projection/View matrix uniform buffer			
		vks::initializers::writeDescriptorSet(descriptorSet, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER, 0, &uniformBuffers.view.descriptor),
		// Binding 1 : Instance matrix as dynamic uniform buffer
		vks::initializers::writeDescriptorSet(descriptorSet, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC, 1, &dynamicBufferInfo),
	};

	vkUpdateDescriptorSets(device, static_cast<uint32_t>(writeDescriptorSets.size()), writeDescriptorSets.data(), 0, NULL);
}
```

**Comment by SaschaWillems** [comment]
The validation message clearly states that you're violating a device limit, so I don't think just setting the range to a higher value is the right solution here.

**Comment by cforfang** [comment]
> I don't think just setting the range to a higher value is the right solution here.

@SaschaWillems wonder if you're misunderstanding the issue :) If you look as the patch (https://github.com/SaschaWillems/Vulkan/commit/db923241322ff9be3faf7e6cbd6ee64ffcc431a7), it effectively changes `dynamicBufferInfo.range` from VK_WHOLE_SIZE (some large value depending on the VkBuffer size) to `sizeof(glm::mat4)` -- the actual amount of data accessed by the shader.

Note again the error: ```VkDescriptorBufferInfo range is VK_WHOLE_SIZE but effective range (320000) is greater than this device's maxUniformBufferRange (65536).```

The problem is the current version of the code is binding the range `[offset, size_of_buffer]` to the shader, and then only accessing the first `sizeof(mat4)` of this. If `size_of_buffer - offset` is larger than `maxUniformBufferRange` this fails.

With the patch the range `[offset, offset + 256]` is bound, for a range of 256, which is within the limit so everything works in all cases.

**Comment by SaschaWillems** [comment]
Okay, then I'll take another look at this.

**Comment by starcatter** [comment]
Seeing as this is still open, and currently my area of study, I'll point out that the issue lies in `vks::VulkanDevice::createBuffer()` and `vks::Buffer::setupDescriptor()` assuming that the buffer's descriptor should always set `VkDescriptorBufferInfo::range` to the size of the buffer being created. 

In case of dynamic uniform buffers, the entire idea is that the underlying buffer is much larger than the uniform data visible to the shader, which only consumes indexed chunks of the memory. The size of that chunk is what the `VkDescriptorBufferInfo::range` field communicates.

Currently it seems the framework for creating these examples doesn't support this in an elegant way, although it's possible to override the range field right after creating the buffer, in `prepareUniformBuffers()`:
after:
```
		// Uniform buffer object with per-object matrices
		VK_CHECK_RESULT(vulkanDevice->createBuffer(
			VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT,
			VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT,
			&uniformBuffers.dynamic,
			bufferSize));
```
add:
```
		uniformBuffers.dynamic.descriptor.range = sizeof(glm::mat4);`
```
Which, while not elegant, does a good job of pointing out that only a small part of the buffer will be visible to the shader.

**Comment by SaschaWillems** [comment]
Fixed via #873
